Update specification for new device flow (#577)
dfinity/internet-identity: docs/internet-identity-spec.adoc

This interface is currently only used by its own frontend. This tight coupling means that this interface may change, even in incompatible ways. We therefore do not have to apply Candid best practices for backward-compatibility (such as using records for arguments and results).

The summary is given by the following Candid interface:
....

> include::../src/internet_identity/internet_identity.did[]

Reviewer: nice

Reviewer 2: Hmm, it prevents being able to update the spec independently of the code (e.g. design a feature in one PR and implement it later) or to keep legacy or convenience stuff (e.g. monitoring) in the implementation that's not in the spec. At least that was the reasoning back then, maybe less of a problem now.

*Authorization*: This request must be sent to the canister with `caller` that is the self-authenticating id derived from any of the public keys of devices associated with the user before this call.

=== The `exit_device_registration_mode` method

Exits device registration mode immediately. Any non verified tentative devices are discarded.

*Authorization*: This request must be sent to the canister with `caller` that is the self-authenticating id derived from any of the public keys of devices associated with the user before this call.

=== The `add_tentative_device` method

Tentatively adds a new device to the supplied identity anchor and returns a verification code. This code has to be used with the `verify_tentative_device` method to verify this device. If the flow is aborted or not completed within 15 minutes, the tentative device is discarded.

Tentatively added devices cannot be used to login into the management view or authorize authentications for other dApps.

*Authorization*: Anyone can call this

=== The `verify_tentative_device` method

> This method, if called with a valid verification code, adds the tentative device to the user permanently and exits registration mode.

Reviewer: ```suggestion
For an anchor in device registration mode: if called with a valid verification code, adds the tentative device as a regular device to the anchor and exits registration mode. The registration flow is aborted if this method is called five times with invalid verification codes.

Returns an error if called for a device not in registration mode.
```

*Authorization*: This request must be sent to the canister with `caller` that is the self-authenticating id derived from any of the public keys of devices associated with the user before this call.

=== The `lookup` query method

Fetches all device data associated with a user.

*Authorization*: Anyone can call this

=== The `get_anchor_info` method

> Fetches all data associated with a user including registration mode and tentatively registered devices.

Reviewer: ```suggestion
Fetches all data associated with an anchor including registration mode and tentatively registered devices.
```
